### DISCUSSION TITLE
Can't get communication working - ESP32 RX TX pins

### DISCUSSION THREAD

**Discussion by Knicko83** [dewenni/ESP_Buderus_KM271#15] [Q&A] [answered]
Hi,
I'm using an ESP32 (https://www.makershop.de/plattformen/nodemcu/espressif-esp32-dev-kit-board/) in combination with an original Buderus KM271 and a MAX3232. Connecting the buderus to a serial of my PC I managed to get some hex communication running already.

However, downloaded, compiled and connected with the ESP_Buderus_KM271, the WebUI remains empty. WebUI works, heartbeat works, mqtt works - but using an oscilloscope I cannot see anything happening on the serial uart pins; the TX stays high at 3.3V continuously. Also tried GPIO 1/3 and other pins already, same issue.

What do I need to configure to use this ESP32? According to the layout ( https://i0.wp.com/randomnerdtutorials.com/wp-content/uploads/2018/08/ESP32-DOIT-DEVKIT-V1-Board-Pinout-36-GPIOs-updated.jpg?quality=100&strip=all&ssl=1 )
I've set it to GPIO 16 and 17, i.e. the 3rd internal uart of the ESP32:
#define RXD2 17               // ESP32 RX-pin for KM271 communication, align with hardware
#define TXD2 16               // ESP32 TX-pin for KM271 communication, align with hardware

Any idea?

**Comment by dewenni** [comment]
Can you give us some details about how you wired the ESP to the MAX3232?

Have you tried to cross Tx/Rx between the ESP and MAX3232?

**Reply by dewenni** [reply]
I have checked one of my previous project version, where I also used the original KM271 with a RS232 Adapter.

The config was:
#define RXD2   16    => connected to RX of RS232 Adapter
#define TXD2   17    => connected to TX of RS232 Adapter

but I think I have used a serial crossover cable between the KM271 and the RS232 Adapter.

**Comment by Knicko83** [comment]
Morning, and thanks for the fast reaction.
My setting is the following:
(ESP32)GPIO17/TXD2 (pull-Up 15kOhm) -> (MAX3232)TXD -> (serial cable) -> (KM271-SUBD)PIN2:RXD
(ESP32)GPIO16/RXD2 <- (MAX3232)RXD <- (serial cable) <- (KM271-SUBD)PIN3:TXD
(ESP32)GND <-> (MAX3232)GND <-> (serial cable) <-> (KM271-SUBD)PIN5:GND

Today I attached the oscilloscope whilst being connected to the KM271, and I managed to see some communication going on. ESP is shifting levels on TXD2, and soon after some respose comes on RXD2.
However, the Web-Interface / MQTT still remains completely empty from any heating data.

I then switched to a serial debugging ESP32 with the same HW wiring and tried to trigger communication from a terminal according to ( https://www.mikrocontroller.net/attachment/63563/km271-protokoll.txt). Green outgoing / yellow incoming response.

![grafik](https://user-images.githubusercontent.com/30951261/215324661-dbcbae79-c0e8-4140-8c4a-036d4c87a8e1.png)

Basically, I can see that the KM271 is sending 0x02 and 0x10 all the time (i.e. cyclically), regardless whether I send something.
When I send "0xEE 0x00 0x00 0x10 0x03 0xFD" it immediately responds with 0x02 0x10; when I send 0x10 or 0x02 it feels like nothing happens until the next cyclic 0x02 0x10 idle.

--> Some communcation seems to be going on, and the fact that I receive the 0x02 0x10 messages tells me that RX/TX direction is not falsely swapped throughout the whole communication chain
--> My feeling is that the KM271 ignores or misinterpretes any incoming message and never sends out any status information.
--> My Buderus is a Logamatic "R21C7 S0" from around ~2000 - should the RS232 communication work with it, or is it maybe too old or has the wrong internal firmware?
--> Is there any internal setting in the Logamatic that has to be made in order to activate the serial communication?

Any response highly welcome, I wanna get this working :P
And it goes without saying: I really appreciate your work dewenni & the78mole!

**Comment by dewenni** [comment]
> --> Some communcation seems to be going on, and the fact that I receive the 0x02 0x10 messages tells me that RX/TX direction is not falsely swapped throughout the whole communication chain
yes, seems like it is working in general. At least the receive.

>--> My feeling is that the KM271 ignores or misinterpretes any incoming message and never sends out any status information.

Maybe you can try to remove the Pull-Up in at the TX-Pin. In my setup I haven't used an external Pull-Up.

>--> My Buderus is a Logamatic "R21C7 S0" from around ~2000 - should the RS232 communication work with it, or is it maybe too old or has the wrong internal firmware?

I don´t think so. I have also a R2107 S0. 

>--> Is there any internal setting in the Logamatic that has to be made in order to activate the serial communication?

As far as I know there is nothing more to do than to send the "log command". After that, the Logamatic starts communication.

**Comment by Knicko83** [comment, marked as answer]
Hi,
oh dear, that was serious failure analysis over some days. Impossible without an oscilloscope.

First things first for any future readers: the ESP is now working with the ESP via Pins 16/17 (Serial 2), no pullups/pulldowns, serial-cross cable with pins swapped. My Buderus is a logamatic RS2107 S0 with software version 315.

So, what was the problem?
Apparently, my MAX3232 has an internal chip problem, or the chip is cheap fake (bought a readymade board via amazon). There was massive crosstalk from the RS232 RX back to the TX channel, which we could clearly identify to lie within the MAX3232 chip. The ESP was always requesting the appropriate bitsequence 0xEE 0x0 0x0 0x10 0x3 0xFD (yellow) to enable logmode after receiving the 0x2 0x10 trigger from buderus (blue), but the buderus never had a chance to understand this mess.

![20230130_162843](https://user-images.githubusercontent.com/30951261/216574609-63d54758-ec1e-4f9d-bd20-6f9a2867d455.jpg)

I really couldn't believe it when we found out. Found very few people on the internet talking about similar issues - I couldn't get it fixed, even with strong power supply for the internal charge pumps as suggested in some posts. So desoldering and replacing the MAX3232 with a pin-identical St3232cd finally made it work.

**Reply by dewenni** [reply]
thanks for the feedback and good to hear that it is finally working.
